pkg/connection/store/vault/client/client.go

```diff
@@ -0,0 +1,238 @@
+	if err := k.Get(path, existing); resource.Ignore(IsNotFound, err) != nil {
+		return errors.Wrap(err, errGet)
+	}
+
+	if k.version == v1.VaultKVVersionV1 {
+		dp, changed := dataPayloadV1(existing, secret)
+		if !changed {
+			// No metadata in v1 secrets.
+			// Hence, already up to date.
+			return nil
+		}
+		_, err := k.client.Write(filepath.Join(k.mountPath, path), dp)
+		return errors.Wrap(err, errWriteData)
+	}
+
+	dp, changed := dataPayloadV2(existing, secret)
+	if changed {
+		if _, err := k.client.Write(filepath.Join(k.mountPath, "data", path), dp); err != nil {
+			return errors.Wrap(err, errWriteData)
+		}
+	}
+
+	mp, changed := metadataPayload(existing.CustomMeta, secret.CustomMeta)
+	// Update metadata only if there is some Data in secret.
+	if len(existing.Data) > 0 && changed {
+		if _, err := k.client.Write(filepath.Join(k.mountPath, "metadata", path), mp); err != nil {
+			return errors.Wrap(err, errWriteMetadata)
+		}
+	}
+
+	return nil
+}
+
+// Delete deletes KVSecret at the given path.
+func (k *KVAdditiveClient) Delete(path string) error {
+	if k.version == v1.VaultKVVersionV1 {
+		_, err := k.client.Delete(filepath.Join(k.mountPath, path))
+		return errors.Wrap(err, errDelete)
+	}
+
+	// Note(turkenh): With KVAdditiveClient v2, we need to delete metadata and all versions:
+	// https://www.vaultproject.io/api-docs/secret/kv/kv-v2#delete-metadata-and-all-versions
+	_, err := k.client.Delete(filepath.Join(k.mountPath, "metadata", path))
+	return errors.Wrap(err, errDelete)
+}
+
+func dataPayloadV2(existing, new *KVSecret) (map[string]interface{}, bool) {
+	data := make(map[string]interface{}, len(existing.Data)+len(new.Data))
+	for k, v := range existing.Data {
+		data[k] = v
+	}
+	changed := false
+	for k, v := range new.Data {
+		if ev, ok := existing.Data[k]; !ok || ev != v {
+			changed = true
+			data[k] = v
+		}
+	}
+	ver := json.Number("0")
+	if existing.version != "" {
+		ver = existing.version
+	}
+	return map[string]interface{}{
+		"options": map[string]interface{}{
+			"cas": ver,
+		},
+		"data": data,
+	}, changed
+}
+
+func metadataPayload(existing, new map[string]interface{}) (map[string]interface{}, bool) {
+	payload := map[string]interface{}{
+		"custom_metadata": new,
+	}
+	if len(existing) != len(new) {
+		return payload, true
+	}
+	for k, v := range new {
+		if ev, ok := existing[k]; !ok || ev != v {
+			return payload, true
+		}
+	}
+	return payload, false
+}
+
+func dataPayloadV1(existing, new *KVSecret) (map[string]interface{}, bool) {
+	data := make(map[string]interface{}, len(existing.Data)+len(new.Data))
+	for k, v := range existing.Data {
+		data[k] = v
+	}
+	changed := false
+	for k, v := range new.Data {
+		if ev, ok := existing.Data[k]; !ok || ev != v {
+			changed = true
+			data[k] = v
+		}
+	}
+	return data, changed
+}
+
+func (k *KVAdditiveClient) parseAsKVSecret(s *api.Secret, kv *KVSecret) error {
+	if k.version == v1.VaultKVVersionV1 {
+		kv.Data = s.Data
+	}
+
+	// kv version is v2
```

Comment:
Is this true? The clause above makes me think it could be v1 too.

Reply:
We should have returned at the end of the first if clause. So the comment is true indeed.
Fixed!